paddle/operators/ctc_edit_distance_op.cc

```diff
@@ -0,0 +1,87 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#include "paddle/operators/ctc_edit_distance_op.h"
+
+namespace paddle {
+namespace operators {
+
+class CTCEditDistanceOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("X1"), "Input(X1) shouldn't be null.");
+    PADDLE_ENFORCE(ctx->HasInput("X2"), "Input(X2) shouldn't be null.");
+    PADDLE_ENFORCE(ctx->HasOutput("Out"), "Output(Out) shouldn't be null.");
+    ctx->SetOutputDim("Out", {1});
+  }
+
+ protected:
+  framework::OpKernelType GetKernelType(
+      const framework::ExecutionContext &ctx) const override {
+    return framework::OpKernelType(framework::DataType::FP32,
+                                   ctx.device_context());
+  }
+};
+
+class CTCEditDistanceOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  CTCEditDistanceOpMaker(framework::OpProto *proto,
+                         framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X1",
+             "(2-D tensor with shape [M x 1]) The indices for "
+             "hypothesis string");
+    AddInput("X2",
+             "(2-D tensor with shape [N x 1]) The indices "
+             "for reference string.");
+    AddAttr<bool>("normalized",
+                  "(bool, default false) Indicated whether "
+                  "normalize the Output(Out) by the length of reference "
+                  "string (X2).")
+        .SetDefault(false);
```

Comment:
In ctc, the `blank` should be removed from the predicted result. In attention, the `eos` and `sos` should be removed. Maybe here, we need an attribute named `uncare` of type `std::vector<int>`, to set the elements that need to be removed.
It is not implemented in [CTCErrorEvaluator.cpp](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/evaluators/CTCErrorEvaluator.cpp), but is a real need from ocr.

Reply:
I agree with that the operator needs to accept some uncared input tokens. But it would make the code dirty allowing for the CUDA kernel. How about implementing an independent op to remove the uncared tokens?